### ISSUE TITLE
BETA BLOG - Jakarta CDI 4.1

### ISSUE THREAD

**Issue by Azquelt** [OpenLiberty/open-liberty#28201]
The information you provide here will be included in the Open Liberty beta blog post ([example](https://openliberty.io/blog/2020/08/05/jakarta-grpc-beta-20009.html)), which will be published on [openliberty.io/blog/](https://www.openliberty.io/blog/), and potentially elsewhere, to promote this beta feature/function of Open Liberty. For this post to be included in the beta issue please make sure that this is completed by the end of Friday following the GM (Tuesday). The beta and release blogs are created using automation and rely on you following the template's structure. **DO NOT REMOVE/ALTER THE `<GHA>` TAGS THROUGHOUT THIS TEMPLATE.**

Please provide the following information:
&#8203;<GHA-BLOG-RELATED-FEATURES>
1. Which Liberty feature(s) does your update relate to? 
    
    Human-readable name (eg WebSockets feature): Jakarta Contexts and Dependency Injection 4.1
   
    Short feature name (eg websockets-1.0):  `cdi-4.1`
   
   
    </GHA-BLOG-RELATED-FEATURES>

    <GHA-BLOG-TARGET-PERSONA>
2. Who is the target persona? Who do you expect to use the update? eg application developer, operations.  
    
    Application developers.
    
    </GHA-BLOG-TARGET-PERSONA>
   
    <GHA-BLOG-SUMMARY>
3. Provide a summary of the update, including the following points:

CDI allows objects to be bound to lifecycle contexts, to be injected, to be associated with interceptors and decorators, and to interact in a loosely coupled fashion by firing and observing events. Highlights of CDI 4.1 include the ability for interceptors to inspect the annotations used to bind them. This update also brings in new Jakarta EE 11 versions of the Jakarta Annotations and Jakarta Interceptors APIs.

### Retrieve interceptor binding info

Support has been added to allow interceptors to retrieve and inspect the annotations used to bind them. Interceptors can now call [`InvocationContext.getInterceptorBindings()`][get-bindings] or one of the related methods to retrieve the annotations so that they can read values from them. This is particularly useful when some of the fields of an interceptor binding annotation are annotated with `@Nonbinding`. These fields are not used to restrict which interceptor is bound and can now be read by the interceptor when  the interceptor method is called.

For example, a binding annotation for a logging interceptor could look like this:

```java
@InterceptorBinding
public @interface Logged {
    @Nonbinding String value();
}
```

This annotation could be applied to a method on a managed bean like this:
```java
@Logged("myName")
public void myMethod() {
   // ....
}
```

Then an interceptor like this could read the value `"myName"` from the annotation and include it in the log message:
```java
@Interceptor
@Logged("")
public class LoggedInterceptor {

    private static final Logger LOG = Logger.getLogger(LoggedInterceptor.class.getName());

    @AroundInvoke
    public Object intercept(InvocationContext ctx) throws Exception {
        Logged binding = ctx.getInterceptorBinding(Logged.class);
        LOG.info("Method called: " + binding.value());
        return ctx.proceed();
    }
}
```

The new methods added to [`InvocationContext`][invocationContext] make it easier to retrieve the annotations and will return the correct binding annotations whether they're applied to the class or the method and will reflect any changes made to the annotations by a CDI extension.

### Invoker support

Support has been added for extensions to create `Invoker`s which are used to call a method on a CDI managed bean indirectly and can support looking up the object instance and the method arguments from CDI. This feature is intended for people who want to build frameworks which build on or integrate with CDI.

More information on how to create and use an `Invoker` can be found on the CDI spec blog.

### Deprecated methods in BeanManager

`BeanManager.getELResolver` and `BeanManager.wrapExpressionFactory` have been deprecated and will be removed in the next major version so that the CDI API won't need to depend on the Expression Language API. The same methods have been added to `ELAwareBeanManager` which extends `BeanManager` and lives in a new API jar.

These methods were mostly used by EL implementations, but any users who wish to continue using these methods will have to switch to using `ELAwareBeanManager` and include the new API artifact in their dependencies. e.g. by adding it to their maven pom.xml:

```xml
<dependency>
    <groupId>jakarta.enterprise</groupId>
    <artifactId>jakarta.enterprise.cdi-el-api</artifactId>
    <scope>provided</provided>
    <version>4.1.0</version>
</dependency>
```

NOTE: in this beta, `ELAwareBeanManager` cannot be injected directly, but a `BeanManager` can be injected and then cast to an `ELAwareBeanManager`. This will be fixed before release.

### Remove references to Managed Beans specification

Jakarta EE 11 will not include the Managed Beans specification (which was separate from the CDI specification but did many of the same things) and so the associated `@ManagedBean` annotation is removed from the Jakarta Annotations API.

### Changes to EE integration

The parts of the CDI specification which define how CDI integrates with other Jakarta EE technologies have been moved from the CDI specification into the Jakarta EE 11 Platform, Web Profile and Core Profile specifications. This has no effect on end users, but if you use the [CDI specification][cdi-spec] as a documentation reference, you will find that some parts of it are now in the [Plaform][platform-spec], [Web Profile][web-profile-spec] and [Core Profile][core-profile-spec] specifications instead.

To use the CDI 4.1 feature, add `cdi-4.1` to your `server.xml`:

```xml
<featureManager>
  <feature>cdi-4.0</feature>
</featureManager>
```

Further references:
* [CDI API](https://jakarta.ee/specifications/cdi/4.1/apidocs/)
* [Interceptors API](https://jakarta.ee/specifications/interceptors/2.2/apidocs/)
* [Jakarta Annotations API](https://jakarta.ee/specifications/annotations/3.0/apidocs/jakarta.annotation/module-summary.html)
* [CDI Specification](https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1)
* [Interceptors Specification](https://jakarta.ee/specifications/interceptors/2.2/jakarta-interceptors-spec-2.2)
* [Jakarta Annotations Specification](https://jakarta.ee/specifications/annotations/3.0/annotations-spec-3.0)

[get-bindings]: https://jakarta.ee/specifications/interceptors/2.2/apidocs/jakarta.interceptor/jakarta/interceptor/invocationcontext#getInterceptorBindings()
[invocationContext]: https://jakarta.ee/specifications/interceptors/2.2/apidocs/jakarta.interceptor/jakarta/interceptor/invocationcontext
[spec-blog]: https://jakartaee.github.io/cdi/2024/02/27/whats-new-in-cdi41.html
[cdi-spec]: https://jakarta.ee/specifications/cdi/4.1/jakarta-cdi-spec-4.1
[platform-spec]: https://jakarta.ee/specifications/platform/11/
[web-profile-spec]: https://jakarta.ee/specifications/webprofile/11/
[core-profile-spec]: https://jakarta.ee/specifications/coreprofile/11/
    </GHA-BLOG-SUMMARY>

## What happens next?
- Add the label for the beta you're targeting: `target:YY00X-beta`.
- Make sure this blog post is linked back to the Epic for this feature/function.
- Your paragraph will be included in the beta blog post. It might be edited for style and consistency.
- You will be asked to review a draft before publication.
    - Once you've approved the code review, close this issue. 
- If you would _also_ like to write a standalone blog post about your update (highly recommended), raise an issue on the [Open Liberty blogs repo](https://github.com/OpenLiberty/blogs/issues/new/choose). State in the issue that the blog post relates to a specific release so that we can ensure it is published on an appropriate date (it won't be the same day as the beta blog post).

**Comment by Emily-Jiang** [comment]
I have a few comments on this blog:
1. please use heading to list the changes:`
      a. Retrieve InterceptorBinding info
      b. Invoker support
      c. Deprecated methods in BeanManager.
      d. Remove the reference to ManagedBeans
      e. Remove the EE intertegration
2. Jakarta 11 -> Jakarta EE 11.
3. The integration with other EE components were moved to Jakarta EE 11 Platform, Web Profile or Core Profile accordingly. please note not all of them go to the platform spec.

**Comment by Azquelt** [comment]
Thanks @Emily-Jiang, I've made all those changes.

I used "Changes to EE integration" for the last heading because although EE integration has been removed from the spec, it hasn't been removed from Open Liberty and I didn't want the heading to look scary when users will be unaffected.

**Comment by Emily-Jiang** [comment]
Looks great! Thank you @Azquelt for the update!

**Comment by Azquelt** [comment]
This was published as part of the 24.0.0.5 beta blog: https://openliberty.io/blog/2024/05/07/24.0.0.5-beta.html#cdi
